{
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/PDBeurope/pdbe-notebooks/blob/main/variants_embl-ebi_july2025/01_structures_availability.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "**<H1>Structures availability</H1>**\n",
        "**Using PDBe-KB & 3D-Beacons to gain understanding on genetic variants**\n",
        "<img src=\"https://www.ebi.ac.uk/pdbe/docs_dev/logos/images/RGB/PDBe-logo-RGB_2013.png\" height=\"300\" align=\"right\">\n",
        "\n",
        "#Welcome to this notebook!\n",
        "\n",
        "**To use this notebook:**\n",
        "* you will need to have a Google account\n",
        "* you will need to be logged-in to your Google account\n",
        "\n",
        "<br>\n",
        "\n",
        "**To access this notebook:**\n",
        "\n",
        "(1) From Github repository (https://github.com/PDBeurope/pdbe-notebooks/)\n",
        ", click on link at top of the page when viewing the file in Github\n",
        "\n",
        "<br>\n",
        "\n",
        "(2) Visit the Colaboratory page:<br>\n",
        "https://colab.research.google.com/<br>\n",
        "and access the following Github repository *via* the interface:<br>\n",
        "https://github.com/PDBeurope/pdbe-notebooks/\n",
        "\n",
        "--or--\n",
        "\n",
        "upload the file into the Colaboratory interface\n",
        "\n",
        "(for example, after downloading from:\n",
        "https://github.com/PDBeurope/pdbe-notebooks/)"
      ],
      "metadata": {
        "id": "QXp8HG3aOXzS"
      },
      "id": "QXp8HG3aOXzS"
    },
    {
      "cell_type": "markdown",
      "source": [
        "\n",
        "\n",
        "---\n",
        "\n",
        "## How to use this notebook\n",
        "1. To run a code cell, click on the cell to select it. You will notice a play button (▶️) on the left side of the cell. Click on the play button or **press Shift+Enter** to run the code in the selected cell.\n",
        "2. The code will start executing, and you will see the output, if any, displayed below the code cell.\n",
        "3. Move to the next code cell and repeat steps 2 and 3 until you have executed all the desired code cells in sequence.\n",
        "4. The currently running step is indicated by a circle with a stop sign next to it.\n",
        "If you need to stop or interrupt the execution of a code cell, you can click on the stop button (■) located next to the play button.\n",
        "\n",
        "*Remember to run the code cells in the correct order, as their execution might depend on variables or functions defined in previous cells. You can modify the code in a code cell and re-run it to see updated results.*\n",
        "\n",
        "<br>\n"
      ],
      "metadata": {
        "id": "6Mg5aemEWkiz"
      },
      "id": "6Mg5aemEWkiz"
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "\n",
        "## Contact us\n",
        "\n",
        "If you experience any bugs please contact pdbehelp@ebi.ac.uk and put \"Help with\" and the title of the notebook in the subject line of the message.\n"
      ],
      "metadata": {
        "id": "38d-k2Tp9sWN"
      },
      "id": "38d-k2Tp9sWN"
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "# Why use this notebook\n",
        "\n",
        "~\n",
        "\n",
        "This interactive Python notebook demonstrates accessing **Protein Data Bank (PDB)**,\n",
        "\n",
        "using some of the tools developed by the **Protein Data Bank in Europe (PDBe)**:\n",
        "\n",
        "<br>\n",
        "\n",
        "- [PDBe Knowledge Base -> pdbe-kb.org](https://www.ebi.ac.uk/pdbe/pdbe-kb/)\n",
        "\n",
        "- [3D Beacons -> 3d-beacons.org](https://www.ebi.ac.uk/pdbe/pdbe-kb/3dbeacons/)\n",
        "\n",
        "<br>\n",
        "\n",
        "<p>An API (Application Programming Interface) is a programmatic way to obtain information.</p>\n",
        "\n",
        "<p>Virtually every webpage you visit is using one or more APIs.</p>\n",
        "\n",
        "<p>We will be tapping into the potential of APIs to more quickly access and assess information.</p>\n",
        "\n",
        "<p> </p>"
      ],
      "metadata": {
        "id": "OVmjjknE0Ntq"
      },
      "id": "OVmjjknE0Ntq"
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "*We will briefly cover:*\n",
        "### **3D-Beacons**\n",
        "\n",
        "<img src=\"https://github.com/PDBeurope/pdbe-notebooks/blob/main/variants_embl-ebi_july2025/images/API_image.png?raw=true\" height=\"300\" align=\"right\">\n",
        "\n",
        "The main purpose of 3D-Beacons is to provide programmatic access from one web location to BOTH:\n",
        "* **predicted / theoretical structures** (from AlphaFold Database & more)\n",
        "* **experimentally-determined structures** (from Protein Data Bank & more)\n",
        "\n",
        "\n",
        "<br>\n",
        "\n",
        "Some example predicted / theoretical model providers:\n",
        "\n",
        "* AlphaFold DB (AFDB)\n",
        "* AlphaFill\n",
        "* ModelArchive\n",
        "* Protein Ensemble Database (PRD)†\n",
        "* SWISS-MODEL\n",
        "\n",
        "Experimentally-determined model providers:\n",
        "\n",
        "* Protein Data Bank (PDB)\n",
        "* Small angle scattering Biological Data Bank (SASBDB)\n",
        "\n",
        "*† Some models have some experimental evidence, others were generated without explicit experimental data. For this data bank each id corresponds to multiple models of the same protein or protein complex and has been developed to capture insight on intrinsically disordered proteins.*\n",
        "\n",
        "<br>\n",
        "\n",
        "**MORE INFORMATION:**\n",
        "\n",
        "Comprehensive list is available here of current [data providers](https://www.ebi.ac.uk/pdbe/pdbe-kb/3dbeacons/guidelines)\n",
        "\n",
        "A link to main webpage:\n",
        "[3d-beacons.org](https://www.ebi.ac.uk/pdbe/pdbe-kb/3dbeacons/)\n",
        "<br>\n",
        "\n",
        "For direct link to 3D-Beacons API interface, visit [3D-Beacons API page](https://www.ebi.ac.uk/pdbe/pdbe-kb/3dbeacons/api/)\n",
        "\n",
        "If you use this resource, please cite:\n",
        "<br>\n",
        "[M. Varadi, S. Nair, I. Sillitoe et al. 3D-Beacons: decreasing the gap between protein sequences and structures through a federated network of protein structure data resources. GigaScience (2022).](https://doi.org/10.1093/gigascience/giac118)"
      ],
      "metadata": {
        "id": "g3VQfvEXI3F7"
      },
      "id": "g3VQfvEXI3F7"
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "*We will briefly cover:*\n",
        "### **Protein Data Bank in Europe Knowledge Base (PDBe-KB)**\n",
        "\n",
        "**PDBe-KB** provides more information and analysis specifically for experimentally-determined structures deposited in the **Protein Data Bank**.\n",
        "\n",
        "**3D Beacons** is being used <i>'behind-the-scenes'</i> on the **PDBe-KB** pages to enable predicted macromolecule models from AlphaFoldDB to be displayed alongside experimental models.\n",
        "\n",
        "**PDBe-KB** provides information on 3D macromolecular structures in the context of other macromolecular structures.\n",
        "\n",
        "With **PDBe-KB** webpages and its APIs, the following information can be accessed:\n",
        "\n",
        "* Identify PDB ids (aka multiple structures) known to contain related data due to one or more protein chains being associated with the same UniProt ID.\n",
        "* Insight on compounds & residues from considering related structures in the context with each other.\n",
        "* Mappings to CATH, EC, GO, InterPro, Pfam, SCOP, etc\n",
        "* and more...\n",
        "\n",
        "\n",
        "**MORE INFORMATION:**\n",
        "\n",
        "A link to main webpage:\n",
        "[pdbe-kb.org](https://www.ebi.ac.uk/pdbe/pdbe-kb/)\n",
        "<br>\n",
        "\n",
        "For direct link to PDBe-KB's Aggregated API, visit [PDBe-KB Aggregated API page](https://www.ebi.ac.uk/pdbe/graph-api/pdbe_doc/)\n",
        "<br>\n",
        "\n",
        "If you use this resource, please cite:\n",
        "<br>\n",
        "[PDBe-KB consortium, PDBe-KB: collaboratively defining the biological context of structural data , Nucleic Acids Research, Database Issue (2022).](https://doi.org/10.1093/nar/gkab988)"
      ],
      "metadata": {
        "id": "zdgXov-aJNYi"
      },
      "id": "zdgXov-aJNYi"
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 1)  Setting up"
      ],
      "metadata": {
        "id": "cLqmWcyS6tRl"
      },
      "id": "cLqmWcyS6tRl"
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "id": "R7fad_36A4PL"
      },
      "outputs": [],
      "source": [
        "from os import ST_APPEND\n",
        "#@title Press shift+enter to start\n",
        "\n",
        "# @markdown This section installs libraries and sets up Python code to extract information from PDBe-KB's APIs.\n",
        "# @markdown <br>\n",
        "# @markdown <br>\n",
        "# @markdown After running this cell, this Notebook will be able to connect 3D-Beacons and PDBe-KB's APIs.\n",
        "# @markdown <br>\n",
        "# @markdown It will create summary tables for some information about the structures associated with a UniProt ID.\n",
        "# @markdown <br>\n",
        "# @markdown <br>\n",
        "# @markdown The information listed below can also be found either on PDBe-KB pages and/or utilizing 3D Beacons.\n",
        "# @markdown <br>\n",
        "\n",
        "\n",
        "######## LIBRARIES\n",
        "\n",
        "import requests, sys, json\n",
        "import pandas as pd\n",
        "from pprint import pprint\n",
        "\n",
        "print(\"Succesfully installed!\")\n",
        "\n",
        "####### 3D BEACONS FUNCTIONS\n",
        "#defining functions for search and summary\n",
        "\n",
        "def _search_3DBeacons(UniProt_ID):\n",
        "    try:\n",
        "        api_url = f\"https://www.ebi.ac.uk/pdbe/pdbe-kb/3dbeacons/api/v2/uniprot/summary/{UniProt_ID}.json\"\n",
        "        response = requests.get(api_url)\n",
        "        response.raise_for_status()\n",
        "        return response.json().get(\"structures\", [])\n",
        "    except requests.exceptions.RequestException as err:\n",
        "        print(\"Double check if the UniProt id is correct.\")\n",
        "        print()\n",
        "        raise Exception(f\"Error fetching data: {err}\")\n",
        "\n",
        "def _create_structure_dataframe(structures):\n",
        "    if structures:\n",
        "        details_list = [\n",
        "            [\n",
        "                s[\"summary\"][\"model_identifier\"],\n",
        "                s[\"summary\"][\"model_category\"],\n",
        "                s[\"summary\"][\"provider\"],\n",
        "                s[\"summary\"][\"uniprot_start\"],\n",
        "                s[\"summary\"][\"uniprot_end\"],\n",
        "                s[\"summary\"][\"coverage\"],\n",
        "            ]\n",
        "            for s in structures\n",
        "        ]\n",
        "        columns = [\"ID\", \"Type\", \"Provider\", \"UniProt start\", \"UniProt end\",\"Coverage\"]\n",
        "        df = pd.DataFrame(details_list, columns=columns)\n",
        "        df[\"Coverage as %\"] = df[\"Coverage\"] * 100\n",
        "        df[\"Coverage\"] = df[\"Coverage\"].round(5)\n",
        "        df[\"Coverage as %\"] = df[\"Coverage as %\"].round(2)\n",
        "        return df\n",
        "    else:\n",
        "        raise Exception(\"No structures found for the given UniProt ID.\")\n",
        "\n",
        "def all_structure_data(UniProt_ID):\n",
        "    try:\n",
        "        structures = _search_3DBeacons(UniProt_ID)\n",
        "        return _create_structure_dataframe(structures)\n",
        "    except Exception as e:\n",
        "        return f\"Error: {e}\"\n",
        "\n",
        "def displaying_3dbeacon_summary(UniProt_ID):\n",
        "    df_3D_beacons = all_structure_data(UniProt_ID)\n",
        "    if type(df_3D_beacons) == str:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "    elif df_3D_beacons is not None:\n",
        "        num_rows = df_3D_beacons.shape[0]\n",
        "        print(f\"There are {num_rows} structures available associated with UniProt ID {UniProt_ID}\")\n",
        "        print()\n",
        "        return df_3D_beacons.head(num_rows)\n",
        "    else:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "\n",
        "def displaying_3dbeacon_exp_summary(UniProt_ID):\n",
        "    df_3D_beacons = all_structure_data(UniProt_ID)\n",
        "    if type(df_3D_beacons) == str:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "    elif df_3D_beacons is not None:\n",
        "        df_3D_beacons2 = df_3D_beacons.drop(df_3D_beacons[df_3D_beacons['Type'] == 'TEMPLATE-BASED'].index)\n",
        "        df_3D_beacons3 = df_3D_beacons2.drop(df_3D_beacons[df_3D_beacons['Type'] == 'CONFORMATIONAL ENSEMBLE'].index)\n",
        "        df_3D_beacons4 = df_3D_beacons3.drop(df_3D_beacons[df_3D_beacons['Type'] == 'AB-INITIO'].index)\n",
        "        df_3D_beacons5 = df_3D_beacons4.reset_index(drop=True)\n",
        "        num_rows = df_3D_beacons5.shape[0]\n",
        "        print(f\"There are {num_rows} different structures, where experimental data was used for the structure, associated with UniProt ID {UniProt_ID}\")\n",
        "        print()\n",
        "        return df_3D_beacons5.head(num_rows)\n",
        "    else:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "\n",
        "def displaying_3dbeacon_exp_summary2(UniProt_ID):\n",
        "    df_3D_beacons = all_structure_data(UniProt_ID)\n",
        "    if type(df_3D_beacons) == str:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "    elif df_3D_beacons is not None:\n",
        "        df_3D_beacons2 = df_3D_beacons[df_3D_beacons['Provider'] == 'PDBe']\n",
        "        df_3D_beacons3 = df_3D_beacons2.reset_index(drop=True)\n",
        "        num_rows = df_3D_beacons3.shape[0]\n",
        "        print(f\"There are {num_rows} different structures, from the Protein Data Bank (PDB), associated with UniProt ID {UniProt_ID}\")\n",
        "        print()\n",
        "        return df_3D_beacons3.head(num_rows)\n",
        "    else:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "\n",
        "def displaying_3dbeacon_exp_summary3(UniProt_ID):\n",
        "    df_3D_beacons = all_structure_data(UniProt_ID)\n",
        "    if type(df_3D_beacons) == str:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "    elif df_3D_beacons is not None:\n",
        "        df_3D_beacons2 = df_3D_beacons.drop(df_3D_beacons[df_3D_beacons['Type'] == 'EXPERIMENTALLY DETERMINED'].index)\n",
        "        df_3D_beacons3 = df_3D_beacons2.drop(df_3D_beacons[df_3D_beacons['Type'] == 'CONFORMATIONAL ENSEMBLE'].index)\n",
        "        df_3D_beacons4 = df_3D_beacons3.reset_index(drop=True)\n",
        "        num_rows = df_3D_beacons4.shape[0]\n",
        "        print(f\"There are {num_rows} different predicted structures, generated using template-based or Ab initio methods, without explicit experimental data, associated with UniProt ID {UniProt_ID}\")\n",
        "        print()\n",
        "        return df_3D_beacons4.head(num_rows)\n",
        "    else:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "\n",
        "####### 3D BEACONS FUNCTIONS\n",
        "#defining functions for reporting urls\n",
        "\n",
        "def links_for_structures(UniProt_ID):\n",
        "    try:\n",
        "        structures = _search_3DBeacons(UniProt_ID)\n",
        "        print(f\"There are {len(structures)} listed associated with UniProt id {UniProt_ID}\")\n",
        "        list_of_lists =[]\n",
        "        for index in range(0, len(structures)):\n",
        "            structure_details = structures[index]['summary']\n",
        "            model_identifier = structure_details['model_identifier']\n",
        "            model_catergory = structure_details['model_category']\n",
        "            provider = structure_details['provider']\n",
        "            model_url = structure_details['model_url']\n",
        "            model_format = structure_details['model_format']\n",
        "            label = f\"Click on url to download the coordinate file - {model_format} format:\"\n",
        "            details_list = {model_identifier : [label, model_url]}\n",
        "            print()\n",
        "            pprint(details_list)\n",
        "            print()\n",
        "    except Exception as e:\n",
        "        return f\"Error: {e}\"\n",
        "\n",
        "def display_urls_for_structures(UniProt_ID):\n",
        "    try:\n",
        "        link_list = links_for_structures(UniProt_ID)\n",
        "    except Exception as e:\n",
        "        return f\"An error occurred: {e}\"\n",
        "\n",
        "######## PDBe-KB FUNCTIONS\n",
        "#defining functions for search and summary for variants\n",
        "\n",
        "def api_structure_PDBeKB(UniProt_ID):\n",
        "    try:\n",
        "        requestURL = f\"https://www.ebi.ac.uk/pdbe/graph-api/uniprot/unipdb/{UniProt_ID}\"\n",
        "        response = requests.get(requestURL)\n",
        "        response.raise_for_status()\n",
        "        return json.loads(response.text)\n",
        "    except requests.exceptions.RequestException as err:\n",
        "        print(f\"Request Error: {err}\")\n",
        "        return None\n",
        "\n",
        "def _parse_variant_data(structure_data, UniProt_ID):\n",
        "\n",
        "    if type(structure_data) == dict:\n",
        "        structure_details = structure_data[UniProt_ID]['data']\n",
        "        amount = len(structure_details)\n",
        "        UniProt_ID = UniProt_ID\n",
        "        UniProt_summary = []\n",
        "        UniProt_summary_list = []\n",
        "        row_list = []\n",
        "\n",
        "        for index in range(0, amount): # PDB ids\n",
        "\n",
        "            pdb_id = structure_details[index]['accession']\n",
        "            accession_item = structure_details[index]\n",
        "\n",
        "            modified_residues = _modified_list(accession_item['residues'])\n",
        "            mutated_residues = _mutated_list(accession_item['residues'])\n",
        "\n",
        "            if modified_residues == None and mutated_residues == None:\n",
        "                pass\n",
        "            elif modified_residues == None and mutated_residues != None:\n",
        "                row_list = []\n",
        "                for index in range(0, len(mutated_residues)):\n",
        "                    mutated_residues[index].update({\"PDB_ID\" : pdb_id})\n",
        "                row_list.append(mutated_residues)\n",
        "            elif modified_residues != None and mutated_residues == None:\n",
        "                row_list = []\n",
        "                for index in range(0, len(modified_residues)):\n",
        "                    modified_residues[index].update({\"PDB_ID\" : pdb_id})\n",
        "                row_list.append(modified_residues)\n",
        "            elif modified_residues != None and mutated_residues != None:\n",
        "                row_list = []\n",
        "                for index in range(0, len(modified_residues)):\n",
        "                    modified_residues[index].update({\"PDB_ID\" : pdb_id})\n",
        "                row_list.append(modified_residues)\n",
        "                for index in range(0, len(mutated_residues)):\n",
        "                    mutated_residues[index].update({\"PDB_ID\" : pdb_id})\n",
        "                row_list.append(mutated_residues)\n",
        "            UniProt_summary_list.extend(row_list)\n",
        "\n",
        "        for number in range(0,len(UniProt_summary_list)):\n",
        "            UniProt_summary += UniProt_summary_list[number]\n",
        "        return UniProt_summary\n",
        "\n",
        "def _modified_list(residues):\n",
        "    modified_list = []  # Initialize the list outside of the loop\n",
        "    for residue_info in residues:\n",
        "        if 'modification' in residue_info:\n",
        "            modified = _modification_info(residue_info)\n",
        "            modified_list.append(modified)  # Append the mutation to the list\n",
        "    if modified_list == []:\n",
        "        return None\n",
        "    else:\n",
        "        return modified_list  # Return the list\n",
        "\n",
        "def _mutated_list(residues):\n",
        "    mutated_list = []  # Initialize the list outside of the loop\n",
        "    for residue_info in residues:\n",
        "        if 'mutation' in residue_info:\n",
        "            mutation = _mutation_info(residue_info)\n",
        "            mutated_list.append(mutation)  # Append the mutation to the list\n",
        "    if mutated_list == []:\n",
        "        return None\n",
        "    else:\n",
        "        return mutated_list  # Return the list\n",
        "\n",
        "def _mutation_info(residue_info):\n",
        "    if residue_info['startIndex'] == residue_info['endIndex']:\n",
        "        residue_position = residue_info['startIndex']\n",
        "        UniProt_amino_acid = residue_info['startCode']\n",
        "        coordinate_amino_acid = residue_info['pdbCode']\n",
        "        change = residue_info['mutationType']\n",
        "        mutated_residue = {'change': change, 'position' : residue_position, 'from_aa' : UniProt_amino_acid, 'to_aa' : coordinate_amino_acid}\n",
        "        return mutated_residue\n",
        "\n",
        "def _modification_info(residue_info):\n",
        "    if residue_info['startIndex'] == residue_info['endIndex']:\n",
        "        residue_position = residue_info['startIndex']\n",
        "        UniProt_amino_acid = residue_info['startCode']\n",
        "        coordinate_amino_acid = residue_info['pdbCode']\n",
        "        change = \"Modification\"\n",
        "        modified_residue = {'change': change, 'position' : residue_position, 'from_aa' : UniProt_amino_acid, 'to_aa' : coordinate_amino_acid}\n",
        "        return modified_residue\n",
        "\n",
        "def variant_summary(UniProt_ID):\n",
        "    try:\n",
        "        structure_data = api_structure_PDBeKB(UniProt_ID)\n",
        "        UniProt_info = _parse_variant_data(structure_data, UniProt_ID)\n",
        "        return UniProt_info\n",
        "    except Exception as e:\n",
        "        print(\"Double check if the UniProt id is correct.\")\n",
        "        print()\n",
        "        return f\"Error: {e}\"\n",
        "\n",
        "def displaying_variant_summary(UniProt_ID):\n",
        "    list_of_dict = variant_summary(UniProt_ID)\n",
        "    if type(list_of_dict) == str:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "    elif list_of_dict== []:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "    elif list_of_dict is not None:\n",
        "        print(f\"This is the output from entering the UniProt ID {UniProt_ID}.\")\n",
        "        df_variants = pd.DataFrame.from_dict(list_of_dict)\n",
        "        df_variants2 = df_variants.sort_values(['position'], na_position='first', ascending=True, key=pd.to_numeric)\n",
        "        df_variants3 = df_variants2.drop_duplicates() # Remove duplicate instance of same conflicts for same PDB ids\n",
        "        df_variants4 = df_variants3.reset_index()\n",
        "        df_variants5 = df_variants4.drop(['index'], axis=1)\n",
        "        num_rows = df_variants5.shape[0]\n",
        "        print()\n",
        "        return df_variants5.head(num_rows)\n",
        "    else:\n",
        "        print(\"Something went wrong. Check the UniProt ID.\")\n",
        "\n",
        "######## PDBe-KB FUNCTIONS\n",
        "#defining functions for search and summary\n",
        "\n",
        "def parse_general_structure_data(structure_data, UniProt_ID):\n",
        "    structure_details = structure_data[UniProt_ID]['data']\n",
        "\n",
        "    accession_id_list = []\n",
        "    UniProt_summary = {}\n",
        "    print(f\"There are {len(structure_details)} experimentally-determined structures associated with UniProt id {UniProt_ID}.\")\n",
        "    print()\n",
        "\n",
        "    for number in range(0,len(structure_details)):\n",
        "        accession_item = structure_details[number]\n",
        "        accession_id = accession_item['accession']\n",
        "        accession_id_list.append(accession_id)\n",
        "\n",
        "        additionalData = accession_item['additionalData']\n",
        "        resolution = additionalData['resolution']\n",
        "        resolution = round(resolution,3)\n",
        "        experiment = additionalData['experiment']\n",
        "\n",
        "        if experiment == 'X-ray diffraction':\n",
        "            concatenated_list = [experiment,resolution]\n",
        "            UniProt_row = {accession_item['accession'] : concatenated_list}\n",
        "        elif experiment == 'Electron Microscopy':\n",
        "            concatenated_list = [experiment,resolution]\n",
        "            UniProt_row = {accession_item['accession'] : concatenated_list}\n",
        "        elif experiment == 'Solution NMR':\n",
        "            concatenated_list = [experiment,None]\n",
        "            UniProt_row = {accession_item['accession'] : concatenated_list}\n",
        "        else:\n",
        "            concatenated_list = [experiment,resolution]\n",
        "            UniProt_row = {accession_item['accession'] : concatenated_list}\n",
        "        UniProt_summary.update(UniProt_row)\n",
        "    #print(UniProt_summary)\n",
        "    return UniProt_summary\n",
        "\n",
        "def general_summary(UniProt_ID):\n",
        "  try:\n",
        "    structure_data = api_structure_PDBeKB(UniProt_ID)\n",
        "    UniProt_info = parse_general_structure_data(structure_data, UniProt_ID)\n",
        "    return(UniProt_info)\n",
        "  except Exception as e:\n",
        "    print(\"Double check if the UniProt id is correct.\")\n",
        "    print()\n",
        "    return f\"Error: {e}\"\n",
        "\n",
        "def display_exp_structure_summary(UniProt_ID):\n",
        "    try:\n",
        "        data = general_summary(UniProt_ID)\n",
        "        if type(data) == str:\n",
        "            print()\n",
        "            print(\"Something has gone wrong. Double-check if the UniProt ID is okay.\")\n",
        "        else:\n",
        "            df1 = pd.DataFrame.from_dict(data, orient=\"index\", columns=['Experimental Method', 'Resolution (in Angstroms)'])\n",
        "            df2 = df1.reset_index()\n",
        "            df3 = df2.rename(columns = {'index':'PDB_ID'})\n",
        "            num_rows = df3.shape[0]\n",
        "            return df3.head(num_rows)\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred: {e}\")\n",
        "\n",
        "def display_exp_structure_summary2(UniProt_ID,resolution):\n",
        "    try:\n",
        "        data = general_summary(UniProt_ID)\n",
        "        if type(data) == str:\n",
        "            print()\n",
        "            print(\"Something has gone wrong. Double-check if the UniProt ID is okay.\")\n",
        "        elif resolution == str:\n",
        "            print()\n",
        "            print(\"Something has gone wrong. Make sure resolution is given as a number with or without decimals.\")\n",
        "        else:\n",
        "            df1 = pd.DataFrame.from_dict(data, orient=\"index\", columns=['Experimental Method', 'Resolution (in Angstroms)'])\n",
        "            df2 = df1[df1['Resolution (in Angstroms)'] <= resolution]\n",
        "            df3 = df2.reset_index()\n",
        "            df4 = df3.rename(columns = {'index':'PDB_ID'})\n",
        "            num_rows = df4.shape[0]\n",
        "            print(f\"There are {num_rows} experimentally-determined structures, with resolution of {resolution} Angstrom or better, associated with UniProt id {UniProt_ID}.\")\n",
        "            return df4.head(num_rows)\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred: {e}\")\n",
        "\n",
        "def comma_pdb_id_list(UniProt_ID):\n",
        "    try:\n",
        "        data = general_summary(UniProt_ID)\n",
        "        if type(data) == str:\n",
        "            print()\n",
        "            print(\"Something has gone wrong. Double-check if the UniProt ID is okay.\")\n",
        "        else:\n",
        "            df1 = pd.DataFrame.from_dict(data, orient=\"index\", columns=['Experimental Method', 'Resolution (in Angstroms)'])\n",
        "            index_as_list = df1.index.tolist()\n",
        "            print(', '.join(map(str, index_as_list)))\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred: {e}\")\n",
        "\n",
        "def comma_pdb_id_list2(UniProt_ID, resolution):\n",
        "    try:\n",
        "        data = general_summary(UniProt_ID)\n",
        "        if type(data) == str:\n",
        "            print()\n",
        "            print(\"Something has gone wrong. Double-check if the UniProt ID is okay.\")\n",
        "        else:\n",
        "            df1 = pd.DataFrame.from_dict(data, orient=\"index\", columns=['Experimental Method', 'Resolution (in Angstroms)'])\n",
        "            df2 = df1[df1['Resolution (in Angstroms)'] <= resolution]\n",
        "            num_rows = df2.shape[0]\n",
        "            print(f\"There are {num_rows} experimentally-determined structures, with resolution of {resolution} Angstrom or better, associated with UniProt id {UniProt_ID}.\")\n",
        "            index_as_list = df2.index.tolist()\n",
        "            print(', '.join(map(str, index_as_list)))\n",
        "    except Exception as e:\n",
        "        print(f\"An error occurred: {e}\")\n",
        "\n",
        "######## PDBe-KB FUNCTION\n",
        "#generate comma-delimited list of nonmodified PDB IDs\n",
        "\n",
        "def nonmodified_PDB_IDs(UniProt_ID):\n",
        "    general = general_summary(UniProt_ID)\n",
        "    variant = variant_summary(UniProt_ID)\n",
        "\n",
        "    PDB_ID_list = []\n",
        "    for item in range (0, len(variant)):\n",
        "        PDBid = variant[item][\"PDB_ID\"]\n",
        "        PDB_ID_list.append(PDBid)\n",
        "    PDB_IDs_with_variant_list = list(dict.fromkeys(PDB_ID_list))\n",
        "    all_PDB_ID = list(general.keys())\n",
        "\n",
        "    s = set(PDB_IDs_with_variant_list)\n",
        "    PDB_IDs_without_variant_list = [x for x in all_PDB_ID if x not in s]\n",
        "    print(f\"Of these there are {len(PDB_IDs_without_variant_list)} experimentally-determined structures without any modifications or discrepancies with UniProt id {UniProt_ID}.\")\n",
        "\n",
        "    print()\n",
        "\n",
        "    PDB_ID_list = []\n",
        "    for item in range (0, len(PDB_IDs_without_variant_list)):\n",
        "        PDBid = PDB_IDs_without_variant_list[item]\n",
        "        PDB_ID_list.append(PDBid)\n",
        "    str_PDB_ID_list = [str(item) for item in PDB_ID_list]\n",
        "    delimiter_comma = \", \"\n",
        "    print(delimiter_comma.join(str_PDB_ID_list))\n",
        "\n",
        "######## GENERAL FUNCTION\n",
        "#converting a json into a comma-delimited list\n",
        "\n",
        "def getting_list_of_ids(load):\n",
        "    ID_list = []\n",
        "    if type(load) == int or type(load) == float:\n",
        "        print(\"Something went wrong. Check input is JSON.\")\n",
        "    elif load == None:\n",
        "        print(\"Something went wrong. Check input is JSON.\")\n",
        "    elif \"PDB_ID\" in list(load[0].keys()):\n",
        "        for item in range (0, len(load)):\n",
        "            PDBid = load[item][\"PDB_ID\"]\n",
        "            ID_list.append(PDBid)\n",
        "        str_PDB_ID_list = [str(item) for item in ID_list]\n",
        "        delimiter_comma = \", \"\n",
        "        print(delimiter_comma.join(str_PDB_ID_list))\n",
        "    elif \"ID\" in list(load[0].keys()):\n",
        "        for item in range (0, len(load)):\n",
        "            id = load[item][\"ID\"]\n",
        "            ID_list.append(id)\n",
        "        str_ID_list = [str(item) for item in ID_list]\n",
        "        delimiter_comma = \", \"\n",
        "        print(delimiter_comma.join(str_ID_list))\n",
        "    else:\n",
        "      print(f\"Examine JSON. Should be from table with column header of either 'ID' or 'PDB_ID'.\")"
      ],
      "id": "R7fad_36A4PL"
    },
    {
      "cell_type": "markdown",
      "source": [
        "# 2)  Structures available using **3D-Beacons**\n",
        "This section reports on predicted and experimentally-determined structures structures for a UniProt ID that are available *via* **3D-Beacons**."
      ],
      "metadata": {
        "id": "ZGUCKZwiNXhn"
      },
      "id": "ZGUCKZwiNXhn"
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "These structures/models have a *'type'* assigned such as:\n",
        "\n",
        "* EXPERIMENTALLY DETERMINED\n",
        "* CONFORMATIONAL ENSEMBLE\n",
        "* TEMPLATE-BASED\n",
        "* AB-INITIO\n",
        "\n",
        "If the *'Type'* is **NOT** 'EXPERIMENTALLY DETERMINED' or 'CONFORMATIONAL ENSEMBLE' then it is a predicted structure.\n",
        "\n",
        "If the *'Type'* is 'CONFORMATIONAL ENSEMBLE' then each result is a set of structures that may or may not have explicit experimental evidence."
      ],
      "metadata": {
        "id": "V4nFBuNEQ7LA"
      },
      "id": "V4nFBuNEQ7LA"
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2.1.  Get predicted structures, as well as experimentally determined structures associated with a UniProt ID\n",
        "#@markdown  In this sub-section you will be able to retrieve BOTH predicted structures, as well as experimentally determined structures associated with an Uniprot accession ID.\n",
        "#@markdown  <br>\n",
        "#@markdown  <br>\n",
        "#@markdown  With 3D-Beacons there are options to search based on sequence and Ensembl id as well.\n",
        "#@markdown  <br>\n",
        "#@markdown  <br>\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#Run this code to perform a API call and load result into a DataFrame\n",
        "displaying_3dbeacon_summary(UniProt_ID)\n"
      ],
      "metadata": {
        "cellView": "form",
        "id": "Md-cOAlsHtpx"
      },
      "execution_count": null,
      "outputs": [],
      "id": "Md-cOAlsHtpx"
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2.2. Get predicted / theoretical structural data only\n",
        "#@markdown  In this sub-section you will be able to retrieve predicted / theoretical structures associated with an Uniprot accession ID.\n",
        "#@markdown  <br>\n",
        "#@markdown  <br>\n",
        "#@markdown  <br>\n",
        "\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#Run this code to perform a API call and load result into a DataFrame\n",
        "displaying_3dbeacon_exp_summary3(UniProt_ID)"
      ],
      "metadata": {
        "id": "JxN9njv-RN5o",
        "cellView": "form"
      },
      "id": "JxN9njv-RN5o",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2.3. Get experimentally-determined structural data associated with a UniProt ID\n",
        "#@markdown  In this sub-section you will be able to retrieve experimentally determined data (including the wwPDB and a curated repository of small angle scattering data and models) associated with an Uniprot accession ID.\n",
        "#@markdown  <br>\n",
        "#@markdown  <br>\n",
        "#@markdown  <br>\n",
        "\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#Run this code to perform a API call and load result into a DataFrame\n",
        "displaying_3dbeacon_exp_summary(UniProt_ID)"
      ],
      "metadata": {
        "id": "Q62C_xQt3ZJe",
        "cellView": "form"
      },
      "id": "Q62C_xQt3ZJe",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2.4. Get experimentally-determined structural data from Protein Data Bank\n",
        "#@markdown  In this sub-section you will be able to retrieve experimentally determined data associated with an Uniprot accession ID for the Protein Data Bank only (aka excluding structures from small angle scattering, etc).\n",
        "#@markdown  <br>\n",
        "#@markdown  <br>\n",
        "#@markdown  <br>\n",
        "\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#Run this code to perform a API call and load result into a DataFrame\n",
        "displaying_3dbeacon_exp_summary2(UniProt_ID)"
      ],
      "metadata": {
        "id": "lXrWEcUHlEBy",
        "cellView": "form"
      },
      "id": "lXrWEcUHlEBy",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 2.5.  Get clickable links to download coordinates\n",
        "#@markdown  In this sub-section you will be able to retrieve urls for predicted and experimentally-determined structures associated with a UniProt ID.\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#Run this code to perform a API call and gets urls and display the urls\n",
        "display_urls_for_structures(UniProt_ID)\n"
      ],
      "metadata": {
        "id": "NiVRjOsYIwDC",
        "cellView": "form"
      },
      "id": "NiVRjOsYIwDC",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#3)  Structures available using **PDBe-KB**\n",
        "\n",
        "This section reports on the PDB ids for the experimentally-determined structures where at least one of the protein chains in the structure is associated with a UniProt ID.\n",
        "\n",
        "The identity of these structures as well as more analysis of structures are available *via* **PDBe-KB**."
      ],
      "metadata": {
        "id": "qncyCaFMOzsb"
      },
      "id": "qncyCaFMOzsb"
    },
    {
      "cell_type": "markdown",
      "source": [],
      "metadata": {
        "id": "beZuRw5Vwn72"
      },
      "id": "beZuRw5Vwn72"
    },
    {
      "cell_type": "markdown",
      "source": [
        "---\n",
        "*More details:*\n",
        "<p>The experimentaly-determined structures/data available <i>via</i> <strong>PDBe-KB</strong> are associated with UniProt IDs are due to both manual and semi-automated curation.</p>\n",
        "<p>That is a scientific curator is checking sequence alignments, with a feedback loop with the scientist(s) depositing the data confirming the right UniProt IDs are assigned before the data is made public.</p>\n",
        "<p>If there are structures where there was a lot of engineered mutations or where two or more proteins were engineered to make a new polymer chain to gain scientific insight -- these may not be as findable with certain sequence-based search approaches but will be found listed on <strong>PDBe-KB</strong> pages.</p>\n",
        "\n",
        "<p> </p>\n",
        "\n",
        "<p>If the appropriate UniProt IDs are not available at the time of publishing, then after the structures are published, and there is a semi-automated process to assign unmapped structures to the appropriate UniProt ID. There is a synching process that means ultimately all structures associated with a UniProt ID end up listed on both <strong>PDBe-KB</strong> and UniProt pages</p>\n",
        "<p>\n",
        "This mapping between protein structure and UniProt IDs is part of a process called:</p>\n",
        "\n",
        "[Structure Integration with Function, Taxonomy and Sequence (SIFTS)](https://www.ebi.ac.uk/pdbe/docs/sifts/)."
      ],
      "metadata": {
        "id": "SY6enheJGCjj"
      },
      "id": "SY6enheJGCjj"
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 3.1. Get experimentally-determined coordinates associated with a UniProt ID\n",
        "#@markdown  In this sub-section you will be able to retrieve experimentally-determined coordinates where based on sequence and reported taxonomy there is one or more chains associated with the same specified Uniprot accession ID.\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#Run this code to perform a API call and load result into a DataFrame\n",
        "display_exp_structure_summary(UniProt_ID)"
      ],
      "metadata": {
        "id": "GlTde4OHtu0A",
        "cellView": "form"
      },
      "id": "GlTde4OHtu0A",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 3.2. Press shift+enter to generate a list for copying-and-pasting\n",
        "#@markdown Here is a copy-and-paste comma list of all the PDB ids for one UniProt id.\n",
        "#@markdown <br>\n",
        "#@markdown <br>\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#Run this code to perform a API call and shows results as comma-delimited list\n",
        "comma_pdb_id_list(UniProt_ID)"
      ],
      "metadata": {
        "id": "v-t0VJIugq_q",
        "cellView": "form"
      },
      "id": "v-t0VJIugq_q",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "<img src=\"https://github.com/PDBeurope/pdbe-notebooks/blob/main/variants_embl-ebi_july2025/images/Mol_Viewer_Loading_Multiples.png?raw=true\" align=\"right\">\n",
        "\n",
        "The above list can be pasted into the [Molstar viewer](https://molstar.org/viewer/) and upon *'Apply'* all the coordinates will be loaded into the viewer."
      ],
      "metadata": {
        "id": "9uSjq-AMmAoT"
      },
      "id": "9uSjq-AMmAoT"
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 3.3. Get experimentally-determined coordinates associated with a UniProt ID of resolution 'X' or better\n",
        "#@markdown  In this sub-section you will be able to retrieve experimentally-determined coordinates where based on sequence and reported taxonomy there is one or more chains associated with the same specified Uniprot accession ID.\n",
        "#@markdown\n",
        "#@markdown  This result is additional filtered based on the resolution of the structures.\n",
        "#@markdown  Resolution values correlates to the level of detail in the experimental data that supports the model.\n",
        "#@markdown  The smaller the number for resolution, the better the level of detail in the experimental data.\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "Resolution =  3 #@param {type:\"number\"}\n",
        "\n",
        "#Run this code to perform a API call and load result into a DataFrame\n",
        "display_exp_structure_summary2(UniProt_ID,Resolution)"
      ],
      "metadata": {
        "id": "8jmeuVCnQGGB",
        "cellView": "form"
      },
      "id": "8jmeuVCnQGGB",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 3.4. Press shift+enter to generate a list for copying-and-pasting\n",
        "#@markdown Here is a copy-and-paste comma list of all the PDB ids for one UniProt id.\n",
        "#@markdown <br>\n",
        "#@markdown <br>\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "Resolution =  3 #@param {type:\"number\"}\n",
        "\n",
        "#Run this code to perform a API call and shows results as comma-delimited list\n",
        "comma_pdb_id_list2(UniProt_ID,Resolution)"
      ],
      "metadata": {
        "cellView": "form",
        "id": "7_j_8Yp6xqFx"
      },
      "execution_count": null,
      "outputs": [],
      "id": "7_j_8Yp6xqFx"
    },
    {
      "cell_type": "markdown",
      "source": [
        "#4)  Variant structures available using **PDBe-KB**\n",
        "\n",
        "This section reports on the PDB ids for the experimentally-determined structures with a focus on the presence of protein sequence variants.\n",
        "\n",
        "Structures of variants / engineered mutations of proteins are some times part of research studies. This section provides options to explores whether these exist and what they are."
      ],
      "metadata": {
        "id": "3Udd2aVHulz1"
      },
      "id": "3Udd2aVHulz1"
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 4.1. Get summary of sequence discrepancy (*e.g.* variants) compared to the UniProt sequence\n",
        "#@markdown In this sub-section the variants, engineered mutations, modified residues, etc for experimental structures will be reported.\n",
        "#@markdown <br>\n",
        "#@markdown <br>\n",
        "#@markdown These have been identified as positions that are in conflict with the UniProt sequence.\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#@markdown <br>\n",
        "#@markdown Use filter button and imput the residue number to see if there is an experimental structure for a particular genetic variant.\n",
        "#@markdown\n",
        "#@markdown <br>\n",
        "#@markdown <br>\n",
        "\n",
        "displaying_variant_summary(UniProt_ID)"
      ],
      "metadata": {
        "id": "UIbBveXuS2Go",
        "cellView": "form"
      },
      "execution_count": null,
      "outputs": [],
      "id": "UIbBveXuS2Go"
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 4.2. Press shift+enter to generate a list of PDB ids with no modifications\n",
        "#@markdown Here is a copy-and-paste comma list of the PDB ids with no modifications or discrepancy compared UniProt id.\n",
        "#@markdown <br>\n",
        "#@markdown <br>\n",
        "\n",
        "#Run this code to display the widget\n",
        "UniProt_ID = \"\" #@param {type:\"string\"}\n",
        "\n",
        "#Run this code to perform a API call and shows results as comma-delimited list\n",
        "nonmodified_PDB_IDs(UniProt_ID)"
      ],
      "metadata": {
        "id": "se_9EEpn7uH4",
        "cellView": "form"
      },
      "id": "se_9EEpn7uH4",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "#5)  Convert table output from JSON into a comma-delimited list of IDs\n"
      ],
      "metadata": {
        "id": "3vrWKFleNfE3"
      },
      "id": "3vrWKFleNfE3"
    },
    {
      "cell_type": "code",
      "source": [
        "#@title 5.1. Filter table for subset (*e.g.* variants) and convert from JSON to a comma-delimited list\n",
        "#@markdown In this sub-section the table from any of the sections above can be input as JSON to generate a comma-delimited list of ids.\n",
        "#@markdown <br>\n",
        "#@markdown For example the PDB ids with a specific variant / engineered mutation / modified residues / etc for experimental structures.\n",
        "#@markdown <br>\n",
        "\n",
        "#Run this code to display the widget\n",
        "JSON = None #@param {type:\"raw\"}\n",
        "\n",
        "getting_list_of_ids(JSON)"
      ],
      "metadata": {
        "id": "XEqMMvtIzD2y",
        "cellView": "form"
      },
      "id": "XEqMMvtIzD2y",
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "It should be also noted that there are Cryo Electron Microscopy (CryoEM) maps, where coordinates/models are NOT built, due to limitations of resolution or other reasons. These are not findable using **PDBe-KB** or **3D Beacons** APIs. These Cryo EM maps where coordinates are NOT built do not always have protein sequences listed (so generally less findable from a protein sequence point of view). However, these CryoEM maps may provide insight -- in particular for protein systems where there is lots of flexibility / dynamics and/or few to no experimentally-determined coordinates available.\n",
        "\n",
        "To find CryoEM Maps without coordinates please visit:\n",
        "<br>\n",
        "[Electron Microscopy Data Bank (EMDB)](https://www.ebi.ac.uk/emdb/)"
      ],
      "metadata": {
        "id": "ioUfDNoKjypD"
      },
      "id": "ioUfDNoKjypD"
    },
    {
      "cell_type": "markdown",
      "source": [
        "Copyright 2024 EMBL - European Bioinformatics Institute\n",
        "\n",
        "Licensed under the Apache License, Version 2.0 (the \"License\"); you may not use this file except in compliance with the License. You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0\n",
        "\n",
        "Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on an \"AS IS\" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License."
      ],
      "metadata": {
        "id": "kWp38bpa9Bip"
      },
      "id": "kWp38bpa9Bip"
    }
  ],
  "metadata": {
    "kernelspec": {
      "display_name": "Python 3",
      "name": "python3"
    },
    "language_info": {
      "name": "python"
    },
    "colab": {
      "provenance": [],
      "collapsed_sections": [
        "cLqmWcyS6tRl"
      ],
      "include_colab_link": true
    }
  },
  "nbformat": 4,
  "nbformat_minor": 5
}